src/aposteriori/dnn/data_processing/tools.py

```diff
@@ -0,0 +1,158 @@
+              └─.attrs['label'] Three-letter code for the residue.
+
+    Returns
+    -------
+
+    flat_dataset_map: t.List[t.Tuple]
+        List of tuples with the order
+        [... (pdb_code, chain_id, residue_id,  residue_label) ...]
+    """
+    standard_residues = list(standard_amino_acids.values())
+
+    with h5py.File(frame_dataset, "r") as dataset_file:
+        flat_dataset_map = []
+        # Create flattened dataset structure:
+        for pdb_code in dataset_file:
+            for chain_id in dataset_file[pdb_code].keys():
+                for residue_id in dataset_file[pdb_code][chain_id].keys():
+                    residue_label = dataset_file[pdb_code][chain_id][
+                        str(residue_id)
+                    ].attrs["label"]
+
+                    if residue_label in standard_residues:
+                        pass
+                    # If uncommon, attempt conversion of label
+                    elif residue_label in UNCOMMON_RESIDUE_DICT.keys():
+                        print(f"{residue_label} is not a standard residue.")
+                        # Convert residue to common residue
+                        residue_label = UNCOMMON_RESIDUE_DICT[residue_label]
+                        print(f"Residue converted to {residue_label}.")
+                    else:
+                        print("Warning message")
+                        assert (
+                            residue_label in standard_residues
+                        ), f"Expected natural amino acid, but got {residue_label}."
+
+                    flat_dataset_map.append(
+                        (pdb_code, chain_id, residue_id, residue_label)
+                    )
+
+    return flat_dataset_map
+
+
+def balance_dataset(flat_dataset_map: t.List[t.Tuple]):
+    """
+    Balances the dataset by undersampling the least present residue.
+
+    Parameters
+    ----------
+    flat_dataset_map: t.List[t.Tuple]
+        List of tuples with the order
+        [... (pdb_code, chain_id, residue_id,  residue_label) ...]
+
+    Returns
+    -------
+    balanced_dataset_map: t.List[t.Tuple]
+        Balanced list of tuples with the order
+        [... (pdb_code, chain_id, residue_id,  residue_label) ...].
+        This is balanced by undersampling.
+    """
+    # Balancing the dataset:
+    shuffle(flat_dataset_map)
+    standard_residues = sorted(list(standard_amino_acids.values()))
+    # Count all residues in dataset:
+    all_residues_in_dataset = [res[-1] for res in flat_dataset_map]
+    # Count all residues and calculate the maximum number of residue per class:
+    counted_residue_in_dataset = Counter(all_residues_in_dataset)
+    # Count how many residues to extract per class:
+    max_res_num = counted_residue_in_dataset[
+        min(counted_residue_in_dataset, key=counted_residue_in_dataset.get)
+    ]
+    balanced_dataset_map = []
+    # Sort dataset by residue:
+    flat_dataset_map = sorted(flat_dataset_map, key=itemgetter(3))
```

Comment:
So, there are two reasons I've chosen itemgetter: 

1. It tends to outperform lambda when the list gets longer. This is marginal, but I thought that given the amount of proteins and frames, it'd be good to have better performance. 

Source: https://stackoverflow.com/questions/17243620/operator-itemgetter-or-lambda

2. I generally found it more readable at the time of writing so I kind of went for the PEP20

Reply:
Great, that's good to know. I have no preference, but I use anonymous functions quite liberally in my code, so that's the default style I use. Interesting about the perf difference.